crates/walrus-core/src/encoding/quilt_encoding.rs

```diff
@@ -0,0 +1,1442 @@
+            current_col += cols_needed;
+            quilt_index.quilt_patches[i].end_index =
+                u16::try_from(current_col).expect("current_col should fit in u16");
+        }
+
+        let mut final_index_data = Vec::with_capacity(index_total_size);
+        let index_size_u64 = index_total_size as u64;
+        final_index_data.extend_from_slice(QuiltVersionV1::quilt_type_bytes());
+        final_index_data.extend_from_slice(&index_size_u64.to_le_bytes());
+        final_index_data
+            .extend_from_slice(&bcs::to_bytes(&quilt_index).expect("Serialization should succeed"));
+
+        // Add the index to the quilt.
+        add_blob_to_data(&final_index_data, 0);
+
+        Ok(QuiltV1 {
+            data,
+            row_size,
+            quilt_index,
+            symbol_size,
+        })
+    }
+
+    /// Encodes the blobs into a quilt and returns the slivers.
+    fn encode(&self) -> Result<Vec<SliverPair>, QuiltError> {
+        let _guard = self.span.enter();
+        tracing::event!(Level::DEBUG, "starting to encode quilt");
+
+        let quilt = self.construct_quilt()?;
+        let encoder = BlobEncoder::new(self.config.clone(), quilt.data()).map_err(|_| {
+            QuiltError::QuiltOversize(format!("quilt is too large: {}", quilt.data().len()))
+        })?;
+        assert_eq!(encoder.symbol_usize(), quilt.symbol_size());
+        Ok(encoder.encode())
+    }
+
+    /// Encodes the blobs into a quilt and returns the slivers and metadata.
+    fn encode_with_metadata(&self) -> Result<(Vec<SliverPair>, QuiltMetadata), QuiltError> {
+        let _guard = self.span.enter();
+        tracing::event!(Level::DEBUG, "starting to encode quilt with metadata");
+
+        let quilt = self.construct_quilt()?;
+        tracing::event!(
+            Level::DEBUG,
+            "construct quilt success {}",
+            quilt.data().len()
+        );
+        let encoder = BlobEncoder::new(self.config.clone(), quilt.data()).map_err(|_| {
+            QuiltError::QuiltOversize(format!("quilt is too large: {}", quilt.data.len()))
+        })?;
+
+        assert_eq!(encoder.symbol_usize(), quilt.symbol_size);
+
+        let (sliver_pairs, metadata) = encoder.encode_with_metadata();
+        let quilt_metadata = QuiltMetadata::V1(QuiltMetadataV1 {
+            quilt_blob_id: *metadata.blob_id(),
+            metadata: metadata.metadata().clone(),
+            index: QuiltIndexV1 {
+                quilt_patches: quilt.quilt_index().quilt_patches.clone(),
+            },
+        });
+
+        Ok((sliver_pairs, quilt_metadata))
+    }
+}
+
+/// A quilt decoder of version V1.
+#[derive(Debug)]
+pub struct QuiltDecoderV1<'a> {
+    slivers: HashMap<SliverIndex, &'a SliverData<Secondary>>,
+    quilt_index: Option<QuiltIndexV1>,
+}
+
+impl<'a> QuiltDecoderApi<'a, QuiltVersionV1> for QuiltDecoderV1<'a> {
+    fn decode_quilt_index(&mut self) -> Result<&QuiltIndexV1, QuiltError> {
+        let index = SliverIndex(0);
+
```

Comment:
Nit: Can we call this `first_sliver_index` and remove the blank line? (Or just remove the variable.)